Fix another token problem and no unimo query

sebageek opened the pull request:
The first commit should fix the last "token invalid" bug, which causes queries to fail. The second one reduces API load a tiny bit.

swagner-de commented on networking_aci/plugins/ml2/drivers/mech_aci/cobra_client.py:
```diff
@@ -95,7 +102,7 @@ def __init__(self, hosts, user, password, ssl, verify=False, request_timeout=90)
 
     def login(self):
-        # TODO handle multiple hosts
+        # NOTE: login() is called by _retry(). To avoid deadlocks it should not call any methods that trigger a _retry()
```
You mean to say login itself shall not call any function with a `_retry` decorator? Wouldn't we end in a RecursionError then?

sebageek replied:
If the query fails, triggers a login and login would trigger a `_retry()`, then it would wait for the lock that is hold by the same process, thus deadlock.

notandy commented on networking_aci/plugins/ml2/drivers/mech_aci/cobra_client.py:
```diff
@@ -37,6 +38,7 @@
                        rexc.HTTPError, LoginError)
 RETRY_EXCEPTIONS = FALLBACK_EXCEPTIONS + (SSLError, CommitError, QueryError)
 requests.packages.urllib3.disable_warnings()
+_TOKEN_REFRESH_LOCK = threading.Lock()
```
Just a question out of interest, isn't cobra used from a greenthread, so python-thread locking doesn't work here?

sebageek replied:
Yes, cobra is called from a greethread. I wasn't sure when I wrote this code if threading.Lock works with greenthreads, but concluded it should work, as it is used by oslo.concurrency, e.g. [here](https://github.com/openstack/oslo.concurrency/blob/master/oslo_concurrency/lockutils.py#L103).

I just checked in the aci driver backdoor and threading.Lock looks very monkeypatched :)
```
>>> threading.Lock.__module__
'eventlet.green.thread'
>>> threading.Lock.func_code
<code object allocate_lock at 0x7fe070972130, file "/var/lib/openstack/local/lib/python2.7/site-packages/eventlet/green/thread.py", line 78>

```

notandy replied:
what a mess ^^ Thanks for clarification, good to know

sebageek committed:
Call correct method on retry

We need to call wrapper() instead of func(), else it will not understand
the retry parameter and therefore fail to do a retry.

sebageek committed:
Set a default for token refresh time

In some obscure situations refreshTime might be None. In these cases we
just want to trigger a relogin. With a default value of 0 we will be
outside of any token refresh limits and a relogin will be triggered.

sebageek committed:
Avoid race condition in token refresh with locks

When two threads concurrently try to refresh the auth token, then one
might fail. Along with it its query fails. As we don't want that we
protect the refresh with a lock. To avoid two refreshes we check inside
the lock if another thread did already refresh the token and do nothing
if it has.

sebageek committed:
Do not lookup 'uni' dn in ACI

Whenever uni_mo() is called it used to do an API call into the APIC to
determine the 'uni' root, which is always 'uni'. Now this method will
just return the string 'uni', which the cobra ACI library can use as
well, plus we save an API call.

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/cobra_client.py:
```diff
@@ -47,37 +49,50 @@ def wrapper(self, *args, **kwargs):
                 else:
                     # reraise
+                    LOG.exception("Commit failed")
                     raise e
```
Why `raise e` and not just `raise`? As I understood it, `raise` would keep the original traceback while `raise e` would create a new one from this point on.

sebageek replied:
Good point, I'll change that

sebageek replied:
It looks like this only swallows stack in python2, whereas in python3 it doesn't swallow it and also includes the exceptionhandler in the stacktrace. For as long as we're stuck on python2 it still makes sense to change it.
```
try:
    int("foo")
except ValueError as e:
    raise e

python2:
Traceback (most recent call last):
  File "foo.py", line 7, in <module>
    raise e
ValueError: invalid literal for int() with base 10: 'foo'

python3:
Traceback (most recent call last):
  File "foo.py", line 7, in <module>
    raise e
  File "foo.py", line 5, in <module>
    int("foo")
ValueError: invalid literal for int() with base 10: 'foo'

    
try:
    int("foo")
except ValueError:
    raise

python2:
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    int("foo")
ValueError: invalid literal for int() with base 10: 'foo'

python3:
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    int("foo")
ValueError: invalid literal for int() with base 10: 'foo'
```

joker-at-work commented on networking_aci/plugins/ml2/drivers/mech_aci/cobra_client.py:
```diff
@@ -47,37 +49,50 @@ def wrapper(self, *args, **kwargs):
+                if trigger_login:
+                    LOG.info("%s - calling login()", msg)
+                    LOG.debug("Acquiring session refresh lock for login")
+                    with _TOKEN_REFRESH_LOCK:
+                        self.login()
```
Is it safe to login again, if another greenthread already logged in? `login()` doesn't check anything. Just does an unconditional login. Does that invalidate the currently-running requests with the old token? I'd assume it doesn't.

In theory, anything could have happened between our checks for login and acquiring the lock. X other greenthreads could have acquired the log since then. So would it make sense to check for validity again or is that just an optimization and it's best to play safe and login unconditionally?

sebageek replied:
Yes, anything could happen inbetween. The `login()` might happen after another login, but I was not really sure what would be better. We end up in this case when the original login has timed out (which should not happen often due to the frequently running syncloop) or when something goes wrong, e.g. an APIC fails and throws SSL Errors at us. Then we would end up in a situation where the token would still be valid, but the APIC would no longer talk to us. I have not looked into cobra (v3) enough to see if they reset the refreshTime on a connection error and just went with the assumption they don't and the safest way would be to call `login()` agan in all cases where we aren't sure.

I'm not entirely sure that this is the best way, but at the time it felt like the best in terms of risk/work/reward. I am open to suggestions. :)

joker-at-work replied:
Especially the SSL-Error case or failed APIC case makes it a good idea to call `login()`. Thanks.

sebageek committed:
Use `raise` instead of `raise e` in retry for py2

While we're still on python2, `raise e` will swallow the stacktrace.
With Python3 this behaviour changes, as here we will have the normal
stack trace, but it also includes the exception handler.
